openedx_tagging/core/tagging/api.py

```diff
@@ -0,0 +1,104 @@
+""""
+Tagging API
+
+Anyone using the openedx_tagging app should use these APIs instead of creating
+or modifying the models directly, since there might be other related model
+changes that you may not know about.
+
+No permissions/rules are enforced by these methods -- these must be enforced in the views.
+
+Please look at the models.py file for more information about the kinds of data
+are stored in this app.
+"""
+from typing import List, Type
+
+from django.db.models import QuerySet
+from django.utils.translation import gettext_lazy as _
+
+from .models import ObjectTag, Tag, Taxonomy
+
+
+def create_taxonomy(
+    name,
+    description=None,
+    enabled=True,
+    required=False,
+    allow_multiple=False,
+    allow_free_text=False,
+) -> Taxonomy:
+    """
+    Creates, saves, and returns a new Taxonomy with the given attributes.
+    """
+    return Taxonomy.objects.create(
+        name=name,
+        description=description,
+        enabled=enabled,
+        required=required,
+        allow_multiple=allow_multiple,
+        allow_free_text=allow_free_text,
+    )
+
+
+def get_taxonomies(enabled=True) -> QuerySet:
+    """
+    Returns a queryset containing the enabled taxonomies, sorted by name.
+    If you want the disabled taxonomies, pass enabled=False.
+    If you want all taxonomies (both enabled and disabled), pass enabled=None.
+    """
+    queryset = Taxonomy.objects.order_by("name", "id")
+    if enabled is None:
+        return queryset.all()
+    return queryset.filter(enabled=enabled)
+
+
+def get_tags(taxonomy: Taxonomy) -> List[Tag]:
+    """
+    Returns a list of predefined tags for the given taxonomy.
+
+    Note that if the taxonomy allows free-text tags, then the returned list will be empty.
+    """
+    return taxonomy.get_tags()
+
+
+def resync_tags() -> int:
+    """
+    Reconciles all stored ObjectTag entries with any changes made to their associated taxonomies and tags.
+    """
+    num_changed = 0
+    for object_tag in ObjectTag.objects.all().iterator():
```

Comment:
I realize I've been a bit in-and-out of the taxonomy/tagging model discussions, so please pardon me if you folks have already gone over this:

I wonder if this could be less expensive if we make one more layer of model indirection between the Object and the tag. Right now, it seems like we have the global representation of a Tag, and a per-tagged-thing copy in case the global thing goes away. We need that because the Taxonomies+Tags ownership and lifecycle is different from ObjectTags. The process of reconciliation is between per-object copies and the taxonomy's canonical Tag.

But do we really need that many copies? Can't we keep a more normalized representation as long as we create a copy that follows the same ownership as the ObjectTags in question?

What if we kept a copy of the Taxonomy at the layer of the importable/exportable thing (e.g. library, course, learning package)? Take a course for example. Let's say:

* ObjectTags don't have local copies of the name.
* All the ObjectTags for course content is against a course-level Taxonomy.
* Some of those are CourseTaxonomies that are entirely owned and internal to the course.
* Some of those are CourseTaxonomies that serve as links-with-copies to other Taxonomies (CourseTaxonomies controlled by other courses, OrgTaxonomies, SystemTaxonomies).
* Inconsistencies within the Course on import are errors (they can be ignored and the values dropped, or they can stop the process, but there's no reconciliation).
* Reconciliation for updated values happens _between Taxonomies_.

That would significantly cut down on the number of rows that would have to be iterated over during any sort of reconciliation process.

Reply:
Hi @ormsbee , we haven't gone over an option like this, so thank you for raising it! And you're right, it is storage-heavy to store copies of the `name:value` on each ObjectTag, which will store the most rows of any model in this package.

But, this does seem more complicate, so do we care that much about storage space?

> Right now, it seems like we have the global representation of a Tag, and a per-tagged-thing copy in case the global thing goes away.

We have three uses for the `ObjectTag._value` field:

1. What you said ^
2. We can import tagged content without having to have a Taxonomy with these Tags in place before importing.
3. Free-text tags can be anything the content author wants, and are not hierarchical.
   So we don't store them as `Tag`s, we just create an `ObjectTag` with `_value="free text value"` and `tag=None`. This also avoids race conditions; we don't have to worry about creating both a `Tag` and an `ObjectTag` when free-text tagging an object.

(`ObjectTag._name` is also in the model for reasons 1 & 2.)

But there's a lot of dancing going on around this `_value` field, which likely means it's trying to do too much.

If I understand you correctly, you're suggesting something like:

* Taxonomy: canonical source of a taxonomy's tags and metadata. Taxonomy Admins create these, import/export tags.
* ContentTaxonomy: references a `source` Taxonomy`, and keeps a copy of its Tag tree, e.g. as ContentTaxonomyTag
* ContentTag: references a ContentTaxonomy and a ContentTaxonomyTag.
* ContentTaxonomy objects know how to reconcile changes made to its `source` Taxonomy or the source Taxonomy's Tag without losing `ContentTag` data.

So this deals with point #1 above, but how should we handle points #2 and #3?

Reply:
My assumption was that reconciliation would be quite a rare process. If that's the case, I would strongly lean toward the simpler version, even if it uses a bit more space for storage. (Tags are usually quite small so I don't think the space is an issue.)

Even if it's not that rare, I think there's a simpler approach: If it's `ObjectTag.objects.all()` that's the problem, I suggest changing the reconciliation API to work on a single taxonomy or a single course at a time, rather than all tags of all objects. So for example it could take a course ID as its argument, and then do the reconciliation based on `ObjectTag.objects.filter(object_type=..., object_id__startswith="block-v1:the+course+ID@")`. Then you get much more efficient (and targeted) reconciliation at appropriate times (e.g. after a course import), and the data model is simple and easy to reason about.

What about taxonomy import? It seems like you might need to iterate over all the ObjectTags in that case, to see if any of them should be "upgraded" to new values in the taxonomy based on their `name` and `value`. But I think you can actually do a much more targeted iteration by only iterating the `ObjectTag` that have either (taxonomy_id is None) or (taxonomy_id==imported_taxonomy_id and tag_id is None). Which is again much much faster than iterating all the ObjectTags, because I expect that the vast majority of ObjectTags will be filtered out by that criteria (because they're linked to a taxonomy and/or a tag).

Reply:
@bradenmacdonald 

> If that's the case, I would strongly lean toward the simpler version, even if it uses a bit more space for storage. 

That's my thinking too, and hence this PR.. but I am open to other options @ormsbee .

> Even if it's not that rare, I think there's a simpler approach: If it's ObjectTag.objects.all() that's the problem, I suggest changing the reconciliation API to work on a single taxonomy or a single course at a time, rather than all tags of all objects. 

👍 to adding filter options to the [resync_tags](https://github.com/open-craft/openedx-learning/blob/jill/taxonomy-api/openedx_tagging/core/tagging/api.py#L63-L73) method.

> What about taxonomy import? 

Perfect use of this, yes. I'll see if it'll fit into https://github.com/open-craft/openedx-learning/pull/1 , or create a follow-up task.

Thank you!

Reply:
My apologies: I thought I replied to this yesterday but I never submitted my comment.

@bradenmacdonald: Yes, my concern was much more about the number of records that need to be updated (and any potential locking), and not the space usage.

@pomegranited:

> 2. We can import tagged content without having to have a Taxonomy with these Tags in place before importing.

Yeah, what I was proposing assumes that we'd have local Taxonomies/Tags in place for import, because those taxonomies can be locally controlled at the scope the import is happening.

> 3. Free-text tags can be anything the content author wants, and are not hierarchical.
>     So we don't store them as `Tag`s, we just create an `ObjectTag` with `_value="free text value"` and `tag=None`. This also avoids race conditions; we don't have to worry about creating both a `Tag` and an `ObjectTag` when free-text tagging an object.

I'm fuzzy on this part. It doesn't have to block this PR, but don't we have to worry about race conditions where we're making the same free text value ObjectTag for the same object multiple times? It seems like it would benefit from being a separate model altogether, so we could put a uniqueness constraints around the freeform tag and the thing being tagged, independent of a taxonomy?

Reply:
To be clear: I'm fine with the proposal to keep the general shape of the data structures now and do per-course filtering during reconciliation. We can always add in the local-taxonomy-copy thing later if it turns out we want it. The questions I'm asking now are just to improve my understanding.

Reply:
> It seems like it would benefit from being a separate model altogether, so we could put a uniqueness constraints around the freeform tag and the thing being tagged, independent of a taxonomy?

I'll give it some thought.. originally when I was doing the discovery, I thought we'd end up having to subclass all of these base classes, which would lead to lots and lots of classes, e.g. SystemTaxonomy, SystemTaxonomyTag, SystemTaxonomyObjectTag, ...  And so I was trying to reduce the number of classes used at the base here. But now that I've gotten into the ContentTaxonomy implementation, I didn't even need to create a ContentTag class, because all of the ObjectTag validation is handled by the taxonomy. We'll still need a FreeFormTaxonomy of some sort, even if it has no tags, because that model is where the namespace for the object tags are stored.

Unfortunately we're running short on time for this phase 1, and so I'd rather not tweak it too much more until we actually have a frontend that's using it. But if I find that the frontend is having to do a lot of `if freeform then do this else do something else" before calling the backend, then separating them out will make much more sense.